README.md
# lexer

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/lexer)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/lexer/mod.ts)
[![GitHub](https://img.shields.io/github/license/httpland/http-router)](https://github.com/httpland/http-router/blob/main/LICENSE)

[![test](https://github.com/httpland/http-router/actions/workflows/test.yaml/badge.svg)](https://github.com/httpland/http-router/actions/workflows/test.yaml)

Lexical analyzer for JavaScript.

## Features

- Tiny
- Regex ready
- Maximum munch algorithm
- Fast

## Basic usage

Example of simple addition.

```ts
import { Lexer } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

enum TokenType {
  LET = "Let",
  NUMBER = "Number",
  IDENT = "Ident",
  ASSIGN = "Assign",
  PLUS = "Plus",
  SEMICOLON = "Semicolon",
  WS = " ",
}

const lexer = new Lexer({
  [TokenType.LET]: "let",
  [TokenType.NUMBER]: /\d+/,
  [TokenType.IDENT]: /[a-z]+/i,
  [TokenType.ASSIGN]: "=",
  [TokenType.PLUS]: "+",
  [TokenType.SEMICOLON]: ";",
  [TokenType.WS]: { pattern: /[\s\t]+/, ignore: true },
});
const input = `let sum = 100 + 200;`;
const result = lexer.analyze(input);
assertEquals(result, {
  values: [
    { type: "Let", value: "let", offset: 0, column: 0, line: 1 },
    { type: "Ident", value: "sum", offset: 3, column: 3, line: 1 },
    // ...,
    { type: "Semicolon", value: ";", offset: 19, column: 19, line: 1 },
  ],
});
```

## Map of token type and patterns

The lexer is configured by map of token type and a token patterns.

Several expressions are supported for token patterns.

- string
- regex
- above with options

### String patterns

String pattern recognizes strings as they are as patterns.

```ts
import { Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";

const rules: Rules = {
  IMPORT: "import",
  FROM: "from",
  SEMICOLON: ";",
};
```

String patterns can be used for various tokens such as keywords, separators,
etc., and are the most readable and perform the best.

### Regex patterns

The regex pattern defines a regular expression as a `RegExp` object.

```ts
import { Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
const rules: Rules = {
  IDENT: /[a-z]+/i,
  NUMBER: /\d+/,
};
```

Note that regular expressions cannot be defined as string patterns.

```ts
import { Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
const rules: Rules = {
  NOT_REGEX_OBJECT: "\\s+",
};
```

In this case, matching is performed as the string `\s+`.

Also, `RegExp` flags of `g` is restricted to use. This is because it matches
inputs with Left to Right.

If you use the `g` flag, an error is thrown before execution.

### Patterns with options

The two patterns above allow almost any token to be represented.

For more fine-grained control over the generation of tokens, optional patterns
can be defined.

```ts
import { Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";

const rules: Rules = {
  WS: { pattern: /[\s\t]+/, ignore: true },
};
```

| Name    |      Required      | Description                                          |
| ------- | :----------------: | ---------------------------------------------------- |
| pattern | :white_check_mark: | `string` &#124; `RegExp`<br> Token matching pattern. |
| ignore  |                    | `boolean`<br>Whether the token ignore or not.        |

If `ignore` is set to `true`, then even if the pattern matches, it will not be
added to the generated token stream.

## EOF token

The EOF token is automatically added to the end of the token stream.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const { values } = new Lexer({}).analyze("");
assertEquals(values, [{
  type: "EOF",
  value: "",
  offset: 0,
  column: 0,
  line: 1,
}]);
```

The EOF token is automatically added to the end of the token stream.

To customize the type of EOF token, change the `eof` option.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const { values } = new Lexer({}, { eof: "<eof>" }).analyze("");
assertEquals(values, [{
  type: "<eof>",
  value: "",
  offset: 0,
  column: 0,
  line: 1,
}]);
```

If you do not want the EOF token to be added, set the `eof` option to `false`.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const { values } = new Lexer({}, { eof: false }).analyze("");
assertEquals(values, []);
```

## Maximum munch

The maximum munch is used in the matching algorithm.

This identifies the longest matching pattern as the token, **not** in the order
of definition in the token map.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const rules: Rules = {
  A: "A",
  AAA: "AAA",
  AA: "AA",
};
const { values } = new Lexer(rules).analyze("AAAAA");
assertEquals(values, [
  { type: "AAA", value: "AAA", offset: 0, column: 0, line: 1 },
  { type: "AA", value: "AA", offset: 3, column: 3, line: 1 },
]);
```

The maximum munch eliminates the need to worry about definition order and
reduces the potential for many bugs.

The same is true for the regex pattern.

### Pattern priority

String patterns take precedence over regex patterns. If there is a match in the
string pattern, no match is made in the regex pattern.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const rules: Rules = {
  CONST: "const",
  IDENT: /.+/,
};
const { values } = new Lexer(rules).analyze("constant");
assertEquals(values, [
  { type: "CONST", value: "const", offset: 0, column: 0, line: 1 },
  { type: "IDENT", value: "ant", offset: 5, column: 5, line: 1 },
]);
```

Conclusion:

long string > short string >> long regex > short regex

## Lexical analyzing result

The input is converted to a token stream by the lexer.

The result has the following structure:

| Name   | Description                   |
| ------ | ----------------------------- |
| values | `Token[]` <br> Token streams. |

`Token` structure is:

| Name   | Description                                            |
| ------ | ------------------------------------------------------ |
| type   | `string` <br> Defined token type.                      |
| value  | `string`<br> Actual token text value.                  |
| offset | `number`<br> FStart index of the matched in the input. |

If a character not matching the pattern is encountered, it is aborted and `done`
is returned as `false` along with the token stream up to that point.

### Generic token types

The `type` of a token matches the token type defined in the token map.

The `type` of the token is narrowed by generics.

```ts
import { Lexer, Rules } from "https://deno.land/x/lexer@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

const rules: Rules = {
  CONST: "const",
  IDENT: /.+/,
};
const { values } = new Lexer(rules).analyze("constant");
values.forEach(({ type }) => {
  type; // infer "CONST" | "INDENT"
});
```

## API

All APIs can be found in the
[deno doc](https://doc.deno.land/https/deno.land/x/lexer/mod.ts).

## Benchmark

Benchmark script with comparison to several popular schema library is available.

```bash
deno task bench
```

## License

Copyright © 2022-present [TomokiMiyauci](https://github.com/TomokiMiyauci).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# [1.0.0-beta.2](https://github.com/TomokiMiyauci/lexer/compare/1.0.0-beta.1...1.0.0-beta.2) (2022-11-16)


### Features

* change lexer interface, error handling ([2da83ca](https://github.com/TomokiMiyauci/lexer/commit/2da83ca2c966106a93924e29b22ffa4763fbb55a))
* **lexer:** add changing eof token type ([a17737a](https://github.com/TomokiMiyauci/lexer/commit/a17737ab5490503bdd506529ba196c117f12fae2))
* **lexer:** add disabled eof token ([9e83d49](https://github.com/TomokiMiyauci/lexer/commit/9e83d498b8edeb1b7953e2c0fae9c20b86608e62))
* **lexer:** add eof token at end of token stream by default ([d5632a1](https://github.com/TomokiMiyauci/lexer/commit/d5632a1899f95b8c4b78816bc0e923f6cd15660e))
* **lexer:** add lexer option interface and change unknown type ([ad1d462](https://github.com/TomokiMiyauci/lexer/commit/ad1d462f55e17f5f205eef0e6ee999e379520bc4))
* **lexer:** add offset info to token ([72c46c6](https://github.com/TomokiMiyauci/lexer/commit/72c46c6bdc963b49581a972b830bad4c1fae0abe))
* **lexer:** add position info to result token ([826f3a7](https://github.com/TomokiMiyauci/lexer/commit/826f3a7f1eef76361fd72e9b3b95f09d97a8680b))
* **lexer:** change default unknown type ([d12119e](https://github.com/TomokiMiyauci/lexer/commit/d12119ed2d055bc8197b981983d98383063c243b))
* **lexer:** rename interface field name `unknownType` to `unknown` ([6836460](https://github.com/TomokiMiyauci/lexer/commit/683646085ecd95efd4ba26e0691ab5accb2a4d51))
* **types:** rename interface and alias ([a00af6a](https://github.com/TomokiMiyauci/lexer/commit/a00af6a7976e2dcaf5e08a1fc35df3a5d12d9c1d))


### Performance Improvements

* **lexer:** improve performance of string patterns ([4867e0e](https://github.com/TomokiMiyauci/lexer/commit/4867e0e9bd507d2ca188cdf20f70efa8df37b787))
* **utils:** change line break matching algorism ([d28e098](https://github.com/TomokiMiyauci/lexer/commit/d28e098023c5bd744a81c7d56358018ab0f3827b))

# 1.0.0-beta.1 (2022-09-02)


### Bug Fixes

* **lexer:** fix to not return EOF token when the lex is failed ([d1c5d90](https://github.com/TomokiMiyauci/lexer/commit/d1c5d90bafdfb793fd17c1b0b49be841aec50a7c))


### Features

* **laxer:** add offset property to each token ([f47eb6b](https://github.com/TomokiMiyauci/lexer/commit/f47eb6b7635275db0e0a7191d76b2211e4a8823d))
* **lexer:** accept offset arg that override initial offset ([c30445c](https://github.com/TomokiMiyauci/lexer/commit/c30445c9f724a1a76b200e02a0243a5ebdccd4d7))
* **lexer:** add asserting global flag of regex ([749d338](https://github.com/TomokiMiyauci/lexer/commit/749d338349542d65bcaebf22b56b5569ea51ffd0))
* **lexer:** add generic types for token type ([661ff11](https://github.com/TomokiMiyauci/lexer/commit/661ff11ffb821a75805e64b42492b50d64137152))
* **lexer:** add generics for token type ([ed7078e](https://github.com/TomokiMiyauci/lexer/commit/ed7078e9cb3ac14701f38d7008fc2c4af31e0c0b))
* **lexer:** add lex rule interface and accpet ignore flag ([f1bc0dd](https://github.com/TomokiMiyauci/lexer/commit/f1bc0dda9f087d45b276d769776a37a482d4cdfa))
* **lexer:** add Lexer class with basic lex logic ([22fc0d1](https://github.com/TomokiMiyauci/lexer/commit/22fc0d112e201b16a4af02f99c3ddf56fe9757e1))
* **lexer:** add merge regex flags ([0adde27](https://github.com/TomokiMiyauci/lexer/commit/0adde2717d6cd33e798733a11ad931755316f969))
* **lexer:** add offset value to return value ([b2b5a89](https://github.com/TomokiMiyauci/lexer/commit/b2b5a89f9b7f93c793ad0801a7677b2531288643))
* **lexer:** add treating EOF match pattern ([d59ef43](https://github.com/TomokiMiyauci/lexer/commit/d59ef43c002f2cb6ec887b9d5a7610d0e3b6cba1))
* **lexer:** change lex function interface ([0f121f4](https://github.com/TomokiMiyauci/lexer/commit/0f121f424394c7784f9cce527b879e1631316596))
* **lexer:** remove generic types for token type ([5272c39](https://github.com/TomokiMiyauci/lexer/commit/5272c39833a672cd422e25782f401673383f0970))


lexer.ts
import {
  $,
  type ExtendArg,
  isFalse,
  isRegExp,
  isString,
  mapValues,
  maxBy,
  prop,
} from "./deps.ts";
import {
  assertRegExpFrag,
  columnLine,
  foldByType,
  uniqueChar,
} from "./utils.ts";
import type { FragmentToken, PatternMap, Rule, Rules, Token } from "./types.ts";
import { CursorImpl } from "./cursor.ts";
import { Pattern } from "./mod.ts";

/** Lexer options. */
export interface LexerOptions {
  /** Unknown token type. If the input stream does not match any grammar, set the token as this type.
   * @default "UNKNOWN"
   */
  readonly unknown?: string;

  /** End of file token. Set the token to eod of token stream as this type.
   * Actual token `value` field will be empty string.
   *
   * - `string` - Set EOF token as this type.
   * - `true` - Set EOF token as default type.
   * - `false` - No set EOF token.
   * @default "EOF"
   */
  readonly eof?: string | boolean;
}

/** Result of lexical analyze. */
export interface AnalyzeResult {
  readonly values: Token[];
}

const DEFAULT_UNKNOWN = "UNKNOWN";
const DEFAULT_EOF = "EOF";

/** Lexer Object.
 *
 * @example
 * ```ts
 * import { Lexer } from "https://deno.land/x/lexer@$VERSION/mod.ts";
 * import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";
 *
 * const lexer = new Lexer({
 *   LET: "let",
 *   NUMBER: /\d+/,
 *   IDENT: /[a-z]+/i,
 *   ASSIGN: "=",
 *   PLUS: "+",
 *   SEMICOLON: ";",
 *   WS: {
 *     pattern: /[\s\t]+/,
 *     ignore: true,
 *   },
 * });
 * const input = `let sum = 100 + 200;`;
 * const result = lexer.analyze(input);
 * assertEquals(result, {
 *   values: [
 *     { type: "Let", value: "let", offset: 0, column: 0, line: 1 },
 *     { type: "Ident", value: "sum", offset: 3, column: 3, line: 1 },
 *     // ...,
 *     { type: "Semicolon", value: ";", offset: 19, column: 19, line: 1 },
 *   ],
 * });
 * ```
 * @throws {Error} When the regex pattern has `g` flag.
 */
export class Lexer {
  #ruleMap: TypePatternContext;
  #unknown: string;
  #eof: string;
  #enableEof: boolean;

  constructor(rules: Rules, options?: LexerOptions) {
    const { eof: _eof = DEFAULT_EOF, unknown = DEFAULT_UNKNOWN } = options ??
      {};
    const eof = isString(_eof) ? _eof : DEFAULT_EOF;
    const detailRules = mapValues(rules, resolvePattern);

    this.#ruleMap = preProcess(detailRules);
    this.#ruleMap.regex.forEach($(prop("pattern"), assertRegExpFrag));
    this.#unknown = unknown;
    this.#eof = eof;
    this.#enableEof = !isFalse(_eof);
  }

  /** Analyze input lexically. */
  analyze(input: string): AnalyzeResult {
    const cursor = new CursorImpl(0);
    const tokens: FragmentToken[] = [];
    const ignoreTypes = new Set<string>();

    while (cursor.current < input.length) {
      const offset = cursor.current;
      const result = resolveRule({
        string: stringResolver,
        regex: regexResolver,
      }, { input, offset: cursor.current, ...this.#ruleMap });

      if (result) {
        if (result.ignore) {
          ignoreTypes.add(result.type);
        }
        tokens.push({ type: result.type, value: result.resolved, offset });
        cursor.next(result.resolved.length);
        continue;
      }

      const current = input.at(offset);
      cursor.next();

      if (current) {
        tokens.push({ type: this.#unknown, value: current, offset });
      }
    }

    if (this.#enableEof) {
      tokens.push({ type: this.#eof, value: "", offset: cursor.current });
    }

    const values = foldByType(columnLine(tokens), this.#unknown).filter(
      isNotIgnoreToken,
    );

    return { values };

    function isNotIgnoreToken(token: Token): boolean {
      return !ignoreTypes.has(token.type);
    }
  }
}

function getTokenByRegex(
  input: string,
  ctx: TypePatternContext["regex"],
): ResolveResult | undefined {
  const tokens: ResolveResult[] = [];

  ctx.forEach(({ pattern, ...rest }) => {
    const result = pattern.exec(input);

    if (result && result[0]) {
      tokens.push({ ...rest, pattern, resolved: result[0] });
    }
  });

  const maybeToken = maxBy(tokens, prop("resolved"));

  return maybeToken;
}

function resolvePattern(pattern: Pattern | Rule): Rule {
  if (isString(pattern) || isRegExp(pattern)) return { pattern };

  return pattern;
}

interface ResolverContext {
  input: string;
  offset: number;
}

function resolveRule(
  resolvers: ResolverMap,
  context: ResolverContext & TypePatternContext,
): ResolveResult | void {
  const r1 = resolvers.string(context, context.string);

  if (r1) return r1;

  const r2 = resolvers.regex(context, context.regex);

  if (r2) return r2;
}

type TypePatternMap = {
  [k in keyof PatternMap]: { pattern: PatternMap[k] } & Rule;
};

type TypePatternContext = {
  [k in keyof TypePatternMap]: (MatchContext & TypePatternMap[k])[];
};

type ResolverMap = {
  [k in keyof TypePatternContext]: ExtendArg<Resolver, TypePatternContext[k]>;
};

interface Typeable {
  type: string;
}

interface MatchContext extends Rule, Typeable {}

interface ResolveResult extends MatchContext {
  resolved: string;
}

interface Resolver {
  (context: ResolverContext): ResolveResult | void;
}

interface PreProcessor {
  (rules: Record<string, Rule>): TypePatternContext;
}

const $pattern = prop("pattern");

const preProcess: PreProcessor = (rules) => {
  const entries = Object.entries(rules);
  const typed = entries.map(([type, value]) => ({ type, ...value }));
  const strings = typed.filter(
    $($pattern, isString),
  ) as (TypePatternMap["string"] & Typeable)[];
  const regexs = typed.filter(
    $($pattern, isRegExp),
  ) as (TypePatternMap["regex"] & Typeable)[];
  const string = strings.toSorted((a, b) =>
    b.pattern.length - a.pattern.length
  );
  const regex = regexs.map((
    { pattern, ...rest },
  ) => ({
    ...rest,
    pattern: new RegExp(pattern, uniqueChar("y", pattern.flags)),
  }));

  return { string, regex };
};

const stringResolver: ResolverMap["string"] = ({ input, offset }, rules) => {
  const characters = input.substring(offset);
  const result = rules.find(({ pattern }) => characters.startsWith(pattern));

  if (result) return { ...result, resolved: result.pattern };
};

const regexResolver: ResolverMap["regex"] = ({ input, offset }, rules) => {
  const regex = rules.map(({ pattern, ...rest }) => {
    pattern.lastIndex = offset;

    return { ...rest, pattern };
  });
  const result = getTokenByRegex(input, regex);

  return result;
};
